LLama/Grammar/GrammarParser.cs

```diff
@@ -0,0 +1,328 @@
+﻿using LLama.Native;
+using System;
+using System.Collections.Generic;
+
+namespace LLama.Grammar
+{
+    /// <summary>
+    /// Source:
+    /// https://github.com/ggerganov/llama.cpp/blob/6381d4e110bd0ec02843a60bbeb8b6fc37a9ace9/common/grammar-parser.cpp
+    /// 
+    /// The commit hash from URL is the actual commit hash that reflects current C# code.
+    /// </summary>
+    internal class GrammarParser
+    {
+        // NOTE: assumes valid utf8 (but checks for overrun)
+        // copied from llama.cpp
+        public Tuple<uint, ReadOnlyMemory<char>> DecodeUTF8(ReadOnlyMemory<char> src)
```

Comment:
A `char` in C++ and a `char` in C# are not equivalent. C# `char` is 16 bits, C++ `char` is (normally) 8 bits. If the original code is using `char` here you probably want to use `byte` instead.

At a higher level, if the purpose of this method is to convert some bytes from UTF8 into some characters you could probably just use `Encoding.UTF8.GetChars` instead.

Reply:
@martindevans Thanks for your notes. I changed it to byte.

> you could probably just use Encoding.UTF8.GetChars instead.

I'd like to keep the C# code structure very similar to the original C++ code. This will make future updates easier, as we can directly compare changes in the C++ code. It also reduces the chance of introducing C#-specific bugs.